src/main/java/org/apache/sling/api/resource/uri/ResourceUri.java

```diff
@@ -0,0 +1,181 @@
+     */
+    public String getHost();
+
+    /**
+     * @return returns the port of the ResourceUri or null if not set
+     */
+    public int getPort();
+
+    /**
+     * @return returns the resource path or null if the URI does not contain a path.
+     */
+    @Override
+    public String getResourcePath();
+
+    /**
+     * @return returns the selector string or null if the URI does not contain selector(s)
+     */
+    @Override
+    public String getSelectorString();
+
+    /**
+     * @return returns the selector array (empty if the URI does not contain selector(s))
+     */
+    @Override
+    public String[] getSelectors();
+
+    /**
+     * @return returns the extension or null if the URI does not contain an extension
+     */
+    @Override
+    public String getExtension();
+
+    /**
+     * @return returns the path parameters or an empty Map if the URI does not contain any
+     */
+    public Map<String, String> getPathParameters();
+
+    /**
+     * @return returns the suffix or null if the URI does not contain a suffix
+     */
+    @Override
+    public String getSuffix();
+
+    /**
+     * @return returns the joint path of resource path, selectors, extension and suffix or null if resource path is not set
+     */
+    public String getPath();
+
+    /**
+     * @return returns the query part of the uri
+     */
+    public String getQuery();
+
+    /**
+     * @return returns the url fragment of the uri
+     */
+    public String getFragment();
+
+    /**
+     * @return scheme specific part of the URI
+     */
+    public String getSchemeSpecificPart();
+
+    /**
+     * @return returns the corresponding suffix resource
+     */
+    @Override
+    public Resource getSuffixResource();
+
+    /**
+     * @return returns true if the uri is either a relative or absolute path (this is the case if scheme and host is empty and the URI path
+     *         is set)
+     */
+    default boolean isPath() {
+        return isBlank(getScheme())
+                && isBlank(getHost())
+                && isNotBlank(getResourcePath());
+    }
+
+    /**
+     * @return true if the uri is a absolute path starting with a slash ('/'). This is the default case for all links to pages and assets in
+     *         AEM.
+     */
+    default boolean isAbsolutePath() {
+        return isPath() && getResourcePath().startsWith(ResourceUriBuilder.CHAR_SLASH);
+    }
+
+    /**
+     * @return true if uri is relative (not an URL and not starting with '/')
+     */
+    default boolean isRelativePath() {
+        return isPath() && !getResourcePath().startsWith(ResourceUriBuilder.CHAR_SLASH);
+    }
+
+    /**
+     * @return true if the uri is an absolute URI containing a scheme.
+     */
+    default boolean isFullUri() {
+        return isNotBlank(getScheme())
+                && isNotBlank(getHost());
+    }
+
+    /**
+     * @return true if the uri is an opaque URI like e.g. mailto:
+     */
+    default boolean isOpaque() {
```

Comment:
This definition contradicts https://docs.oracle.com/javase/8/docs/api/java/net/URI.html#isOpaque--. Even hierarchy based URIs have a scheme specific part.

Reply:
So https://github.com/apache/sling-org-apache-sling-api/blob/4b6c8092e66b20069abd046740a2a0d3ac063fa2/src/test/java/org/apache/sling/api/resource/uri/ResourceUriTest.java#L430 checks each and every URL tested if this method is consistent to URI.isOpaque() (test would fail if that's not the case)

Do you think the example `mailto:` is wrong? https://github.com/apache/sling-org-apache-sling-api/blob/4b6c8092e66b20069abd046740a2a0d3ac063fa2/src/test/java/org/apache/sling/api/resource/uri/ResourceUriTest.java#L297 is testing the mailto case and it should be fine...

Reply:
@kwin Looking closer into it I now see that you meant the default implementation itself - and the root cause for it is that the scheme specific part is held as null for internal state. So this is a good catch, 
b1b44e1 fixes it and https://github.com/apache/sling-org-apache-sling-api/blob/b1b44e1ef161080e4f38cf2b010b389ac18b5055/src/test/java/org/apache/sling/api/resource/uri/ResourceUriTest.java#L417 ensures for each test that URI and resourceUri have the exact same behaviour here.